recipes-core/images/emlinux-image-common.inc

```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) Cybertrust Japan Co., Ltd.
+#
+# SPDX-License-Identifier: MIT
+#
+
+python() {
+    if bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', True, False, d):
+        addpkg='sshd-regen-keys'
+    if bb.utils.contains('IMAGE_INSTALL', 'openssh-server', True, False, d):
+        addpkg='sshd-regen-keys'
+    if bb.utils.contains('IMAGE_INSTALL', 'sshd-regen-keys', True, False, d):
+        addpkg=''
+
+    d.appendVar('IMAGE_INSTALL', ' %s' % addpkg)
```

Comment:
It doesn't have to execute this when  addpkg is empty.

Reply:
Right. How about the following instead?

```
  python() {
      if bb.utils.contains('IMAGE_INSTALL', 'sshd-regen-keys', True, False, d):
          return
      if bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', True, False, d) or \
         bb.utils.contains('IMAGE_INSTALL', 'openssh-server', True, False, d):
          d.appendVar('IMAGE_INSTALL', ' sshd-regen-keys')
  }
```

Reply:
How about this?

```
IMAGE_INSTALL:append = "${@bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', ' sshd-regen-keys', '', d)}"
IMAGE_INSTALL:append = "${@bb.utils.contains('IMAGE_INSTALL', 'openssh-server', ' sshd-regen-keys', '', d)}"
```
or
```
IMAGE_INSTALL:append = "${@ ' sshd-regen-keys' if bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', True, False, d) or bb.utils.contains('IMAGE_INSTALL', 'openssh-server', True, False, d)}"

```

Reply:
> IMAGE_INSTALL:append = "${@bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', ' sshd-regen-keys', '', d)}"
> IMAGE_INSTALL:append = "${@bb.utils.contains('IMAGE_INSTALL', 'openssh-server', ' sshd-regen-keys', '', d)}"

The first one of your suggestions causes a loop error as follows.
```
The variable dependency chain for the failure is: IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> IMAGE_INSTALL -> ROOTFS_PACKAGES -> rootfs_install_pkgs_install

ERROR: Parsing halted due to errors, see error messages above
```

> IMAGE_INSTALL:append = "${@ ' sshd-regen-keys' if bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', True, False, d) or bb.utils.contains('IMAGE_INSTALL', 'openssh-server', True, False, d)}"

The other also fails as well.
```
bb.data_smart.ExpansionError: Failure expanding variable IMAGE_INSTALL, expression was  linux-image-cip ${@'swupdate-certificates' if bb.utils.to_boolean(d.getVar('SWU_SIGNED')) else ''}${@ ' sshd-regen-keys' if bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', True, False, d) or bb.utils.contains('IMAGE_INSTALL', 'openssh-server', True, False, d)} emlinux-customization which triggered exception SyntaxError: expected 'else' after 'if' expression (Var <IMAGE_INSTALL>, line 1)
The variable dependency chain for the failure is: IMAGE_INSTALL -> ROOTFS_PACKAGES -> rootfs_install_pkgs_install

ERROR: Parsing halted due to errors, see error messages above
```

So, the inline expansion causes a loop error, so needs to be implemented as a function.

Reply:
So how about this? (I added "else")

```
IMAGE_INSTALL:append = "${@ ' sshd-regen-keys' if bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', True, False, d) or bb.utils.contains('IMAGE_INSTALL', 'openssh-server', True, False, d) else ''}"
```

Reply:
Got it. Thank you for the confirmation.

Reply:
I've force-pushed. Could you review it again?

> ```
>   python() {
>       if bb.utils.contains('IMAGE_INSTALL', 'sshd-regen-keys', True, False, d):
>           return
>       if bb.utils.contains('IMAGE_PREINSTALL', 'openssh-server', True, False, d) or \
>          bb.utils.contains('IMAGE_INSTALL', 'openssh-server', True, False, d):
>           d.appendVar('IMAGE_INSTALL', ' sshd-regen-keys')
>   }
> ```

Reply:
LGTM